bin/cheritest/cheritest_vm.c

```diff
@@ -602,3 +600,49 @@ cheritest_vm_cow_write(const struct cheri_test *ctp __unused)
 	CHERITEST_CHECK_SYSCALL(close(fd));
 	cheritest_success();
 }
+
+#ifdef __mips__
+
+/*
+ * Test a mapped and unmapped invocation of cloadtags
+ */
+
+void
+test_cloadtags_mapped(const struct cheri_test *ctp __unused)
+{
+	uint64_t tags = 0;
+	void * __capability * __capability c;
+	void * __capability * p;
+
+	p = CHERITEST_CHECK_SYSCALL(mmap(0, PAGE_SIZE, PROT_READ | PROT_WRITE,
+						MAP_ANON, -1, 0));
+	c = (__cheri_tocap void * __capability * __capability) p;
+
+	CHERITEST_VERIFY2(cheri_gettag(c) != 0, "initial cap not constructed");
+
+	p[1] = c;
+	p[2] = c;
+
+	asm volatile ( "cloadtags %[tags], %[ptr]" : [tags]"+r"(tags) : [ptr]"r"(c) );
```

Comment:
```suggestion
	__asm__ ("cloadtags %[tags], %[c]" : [tags]"=r"(tags) : [c]"C"(c) );
```
You don't need `tags` to be a live-in, and can thus drop the zero initialisation. Or you can leave it as `+` but keep the rest of the suggestion if you really do want it. Also doesn't need to be volatile, the inputs/outputs capture the necessary info.

Reply:
This one needs to be `__volatile__`?  Without, clang moves the store to `p[2]` into the branch delay slot for the comparison of `tags` to 6.  Am I breaking an aliasing rule?

Reply:
Oh, you need a memory clobber apparently, it assumes by default you don't dereference things (other than `m` operands which implicitly do an address-of):

> The "memory" clobber tells the compiler that the assembly code performs memory reads or writes to items other than those listed in the input and output operands (for example, accessing the memory pointed to by one of the input parameters).

The alternative would be a dummy `"m"(c)` tacked onto the end (you can't use that itself since it'll include the immediate and not be a register) that you don't actually use. Unfortunately `m` doesn't work with caps in hybrid at the moment... though you do have the plain `p` lying around, so `"m"(p)` should express your dependency properly.

Also, I think, strictly speaking, `volatile` doesn't guarantee that things won't be reordered around it, just that it'll run exactly once? I wouldn't feel comfortable relying on that.